File: src/tools/file-write.ts
// ============================================================================
// file_write 工具——创建或覆盖工作区内文件
// ============================================================================

import * as fs from "node:fs/promises";
import * as path from "node:path";
import { createHash } from "node:crypto";
import type { ITool, ToolResult, ToolExecutionContext, JSONSchema7 } from "./types.js";
import type { PermissionLevel } from "../core/types.js";
import { randomUUID } from "node:crypto";
import { z } from "zod";
import {
  resolveWritableWorkspacePath,
  WorkspacePathError,
} from "./path-policy.js";

// ---------------------------------------------------------------------------
// 输入输出类型
// ---------------------------------------------------------------------------

export interface FileWriteInput {
  path: string;
  content: string;
  createIfMissing?: boolean;
  expectedSha256?: string;
}

const MAX_CONTENT_LENGTH = 2 * 1024 * 1024;
const fileWriteInputValidator = z.object({
  path: z.string().min(1).max(4096),
  content: z.string().max(MAX_CONTENT_LENGTH),
  createIfMissing: z.boolean().optional(),
  expectedSha256: z.string().regex(/^[a-fA-F0-9]{64}$/).optional(),
}).strict();

export interface FileWriteOutput {
  path: string;
  created: boolean;
  preImageSha256: string | null;
  postImageSha256: string;
  unifiedDiff: string;
}

// ---------------------------------------------------------------------------
// 轻量 unified diff 生成器
// ---------------------------------------------------------------------------

function generateUnifiedDiff(
  filePath: string,
  oldContent: string | null,
  newContent: string
): string {
  const oldLines = oldContent ? oldContent.split("\n") : [];
  const newLines = newContent.split("\n");
  const header = oldContent
    ? `--- a/${filePath}\n+++ b/${filePath}\n`
    : `--- /dev/null\n+++ b/${filePath}\n`;

  // 当前实现按整段替换生成可审计差异。
  if (!oldContent) {
    // 新建文件只包含新增行。
    const diffLines = newLines.map((l) => `+${l}`);
    return `${header}@@ -0,0 +1,${newLines.length} @@\n${diffLines.join("\n")}\n`;
  }

  if (oldContent === newContent) {
    return `${header}@@ -1,${oldLines.length} +1,${newLines.length} @@\n (no changes)\n`;
  }

  // 已有文件先寻找公共前后缀，再输出变化区段。
  const lines: string[] = [];
  lines.push(header);
  lines.push(`@@ -1,${oldLines.length} +1,${newLines.length} @@`);

  // 查找公共前缀。
  let commonStart = 0;
  while (
    commonStart < oldLines.length &&
    commonStart < newLines.length &&
    oldLines[commonStart] === newLines[commonStart]
  ) {
    lines.push(` ${oldLines[commonStart]}`);
    commonStart++;
  }

  // 查找公共后缀。
  let oldEnd = oldLines.length - 1;
  let newEnd = newLines.length - 1;
  while (
    oldEnd >= commonStart &&
    newEnd >= commonStart &&
    oldLines[oldEnd] === newLines[newEnd]
  ) {
    oldEnd--;
    newEnd--;
  }

  // 输出变化区段。
  for (let i = commonStart; i <= oldEnd; i++) {
    lines.push(`-${oldLines[i]}`);
  }
  for (let i = commonStart; i <= newEnd; i++) {
    lines.push(`+${newLines[i]}`);
  }

  // 输出公共后缀。
  for (let i = oldEnd + 1; i < oldLines.length; i++) {
    lines.push(` ${oldLines[i]}`);
  }

  return lines.join("\n") + "\n";
}

// ---------------------------------------------------------------------------
// 工具实现
// ---------------------------------------------------------------------------

export class FileWriteTool implements ITool<FileWriteInput, FileWriteOutput> {
  name = "file_write";
  description =
    "创建或覆盖工作区内文件，返回前后哈希和 unified diff。建议先读取文件并提供 expectedSha256。";
  permissionLevel: PermissionLevel = "write-files";
  defaultTimeoutMs = 30_000;
  inputValidator = fileWriteInputValidator;

  inputSchema: JSONSchema7 = {
    type: "object",
    properties: {
      path: {
        type: "string",
        minLength: 1,
        maxLength: 4096,
        description: "要写入的工作区内文件路径",
      },
      content: {
        type: "string",
        maxLength: MAX_CONTENT_LENGTH,
        description: "要写入的新内容",
      },
      createIfMissing: {
        type: "boolean",
        description: "文件不存在时是否创建，默认 true",
      },
      expectedSha256: {
        type: "string",
        pattern: "^[a-fA-F0-9]{64}$",
        description: "当前文件内容的预期 SHA-256，用于防止覆盖并发修改",
      },
    },
    required: ["path", "content"],
    additionalProperties: false,
  };

  idempotencyKey(input: FileWriteInput, _context: ToolExecutionContext): string {
    const contentHash = createHash("sha256").update(input.content).digest("hex");
    return `file_write:${input.path}:${contentHash}`;
  }

  async execute(
    input: FileWriteInput,
    context: ToolExecutionContext
  ): Promise<ToolResult<FileWriteOutput>> {
    const startedAt = new Date().toISOString();
    const createIfMissing = input.createIfMissing !== false; // default true
    let tmpPath: string | undefined;

    try {
      const resolvedPath = await resolveWritableWorkspacePath(input.path, context.cwd);

      // 读取已有内容，用于哈希校验和差异生成。
      let preImageSha256: string | null = null;
      let oldContent: string | null = null;
      let created = false;

      try {
        oldContent = await fs.readFile(resolvedPath, "utf-8");
        preImageSha256 = createHash("sha256").update(oldContent).digest("hex");

        // 写入前检查调用方提供的预期哈希。
        if (input.expectedSha256 && preImageSha256 !== input.expectedSha256) {
          return {
            success: false,
            output: null,
            error: {
              code: "HASH_MISMATCH",
              message: `File hash mismatch. Expected ${input.expectedSha256}, got ${preImageSha256}. The file may have been modified since last read.`,
              retryable: false,
              details: {
                path: input.path,
                expectedSha256: input.expectedSha256,
                actualSha256: preImageSha256,
              },
            },
            durationMs: 0,
            startedAt,
            endedAt: new Date().toISOString(),
            metadata: {},
          };
        }
      } catch (error) {
        if ((error as NodeJS.ErrnoException)?.code !== "ENOENT") throw error;
        // 文件不存在时只在明确允许创建的情况下继续。
        if (!createIfMissing) {
          return {
            success: false,
            output: null,
            error: {
              code: "FILE_NOT_FOUND",
              message: `File does not exist and createIfMissing is false: ${input.path}`,
              retryable: false,
              details: { path: input.path },
            },
            durationMs: 0,
            startedAt,
            endedAt: new Date().toISOString(),
            metadata: {},
          };
        }
        created = true;
      }

      // 确保工作区内的父目录存在。
      await fs.mkdir(path.dirname(resolvedPath), { recursive: true });

      // 先写唯一临时文件，再原子重命名到目标路径。
      tmpPath = `${resolvedPath}.xycli-tmp-${randomUUID()}`;
      await fs.writeFile(tmpPath, input.content, "utf-8");

      try {
        await fs.rename(tmpPath, resolvedPath);
      } catch {
        // 跨设备重命名失败时回退到复制后删除临时文件。
        await fs.copyFile(tmpPath, resolvedPath);
        await fs.unlink(tmpPath);
      }

      const postImageSha256 = createHash("sha256")
        .update(input.content)
        .digest("hex");
      const unifiedDiff = generateUnifiedDiff(input.path, oldContent, input.content);

      const endedAt = new Date().toISOString();
      return {
        success: true,
        output: {
          path: input.path,
          created,
          preImageSha256,
          postImageSha256,
          unifiedDiff,
        },
        error: null,
        durationMs: new Date(endedAt).getTime() - new Date(startedAt).getTime(),
        startedAt,
        endedAt,
        metadata: {
          resolvedPath,
          contentLength: input.content.length,
        },
      };
    } catch (err: unknown) {
      if (tmpPath) {
        try {
          await fs.unlink(tmpPath);
        } catch {
          // 临时文件可能尚未创建或已经完成重命名。
        }
      }
      const endedAt = new Date().toISOString();
      const message = err instanceof Error ? err.message : "Unknown error writing file";
      return {
        success: false,
        output: null,
        error: {
          code: err instanceof WorkspacePathError ? err.code : "FILE_WRITE_ERROR",
          message,
          retryable: false,
          details: { path: input.path },
        },
        durationMs: new Date(endedAt).getTime() - new Date(startedAt).getTime(),
        startedAt,
        endedAt,
        metadata: {},
      };
    }
  }
}

File: crates/xycli-core/src/tools/file_write.rs
//! `file_write`：带并发哈希检查和原子落盘的工作区文件写入。

use std::{path::Path, time::Duration};

use async_trait::async_trait;
use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use tokio::fs;
use uuid::Uuid;

use crate::permission::PermissionLevel;

use super::path_policy::resolve_writable;
use super::{
    Tool, ToolContext, ToolDefinition, ToolResult, object, reject_unknown_fields, required_string,
};

const MAX_CONTENT_LENGTH: usize = 2 * 1024 * 1024;

pub struct FileWriteTool;

fn sha256(content: &[u8]) -> String {
    hex::encode(Sha256::digest(content))
}

fn unified_diff(path: &str, old: Option<&str>, new: &str) -> String {
    let new_lines = new.split('\n').collect::<Vec<_>>();
    let Some(old) = old else {
        let additions = new_lines
            .iter()
            .map(|line| format!("+{line}"))
            .collect::<Vec<_>>()
            .join("\n");
        return format!(
            "--- /dev/null\n+++ b/{path}\n@@ -0,0 +1,{} @@\n{additions}\n",
            new_lines.len()
        );
    };
    let old_lines = old.split('\n').collect::<Vec<_>>();
    if old == new {
        return format!(
            "--- a/{path}\n+++ b/{path}\n@@ -1,{} +1,{} @@\n (no changes)\n",
            old_lines.len(),
            new_lines.len()
        );
    }
    let mut common_start = 0;
    while common_start < old_lines.len()
        && common_start < new_lines.len()
        && old_lines[common_start] == new_lines[common_start]
    {
        common_start += 1;
    }
    let mut old_end = old_lines.len();
    let mut new_end = new_lines.len();
    while old_end > common_start
        && new_end > common_start
        && old_lines[old_end - 1] == new_lines[new_end - 1]
    {
        old_end -= 1;
        new_end -= 1;
    }
    let mut body = Vec::new();
    body.extend(
        old_lines[..common_start]
            .iter()
            .map(|line| format!(" {line}")),
    );
    body.extend(
        old_lines[common_start..old_end]
            .iter()
            .map(|line| format!("-{line}")),
    );
    body.extend(
        new_lines[common_start..new_end]
            .iter()
            .map(|line| format!("+{line}")),
    );
    body.extend(old_lines[old_end..].iter().map(|line| format!(" {line}")));
    format!(
        "--- a/{path}\n+++ b/{path}\n@@ -1,{} +1,{} @@\n{}\n",
        old_lines.len(),
        new_lines.len(),
        body.join("\n")
    )
}

fn valid_hash(value: &str) -> bool {
    value.len() == 64 && value.bytes().all(|byte| byte.is_ascii_hexdigit())
}

#[async_trait]
impl Tool for FileWriteTool {
    fn definition(&self) -> ToolDefinition {
        ToolDefinition {
            name: "file_write",
            description: "创建或覆盖工作区内文件，返回前后哈希和 unified diff。建议先读取文件并提供 expectedSha256。",
            input_schema: json!({
                "type":"object",
                "properties":{
                    "path":{"type":"string","minLength":1,"maxLength":4096},
                    "content":{"type":"string","maxLength":MAX_CONTENT_LENGTH},
                    "createIfMissing":{"type":"boolean"},
                    "expectedSha256":{"type":"string","pattern":"^[a-fA-F0-9]{64}$"}
                },
                "required":["path","content"],
                "additionalProperties":false
            }),
            permission_level: PermissionLevel::WriteFiles,
            default_timeout: Duration::from_secs(30),
        }
    }

    fn validate(&self, input: &Value) -> Result<(), Vec<String>> {
        let map = object(input)?;
        let mut issues = Vec::new();
        reject_unknown_fields(
            map,
            &["path", "content", "createIfMissing", "expectedSha256"],
            &mut issues,
        );
        required_string(map, "path", 4096, &mut issues);
        match map.get("content").and_then(Value::as_str) {
            Some(content) if content.len() <= MAX_CONTENT_LENGTH => {}
            _ => issues.push(format!(
                "content 必须是长度不超过 {MAX_CONTENT_LENGTH} 字节的字符串。"
            )),
        }
        if map
            .get("createIfMissing")
            .is_some_and(|value| !value.is_boolean())
        {
            issues.push("createIfMissing 必须是布尔值。".into());
        }
        if let Some(value) = map.get("expectedSha256") {
            if !value.as_str().is_some_and(valid_hash) {
                issues.push("expectedSha256 必须是 64 位十六进制 SHA-256。".into());
            }
        }
        if issues.is_empty() {
            Ok(())
        } else {
            Err(issues)
        }
    }

    async fn execute(&self, input: Value, context: ToolContext) -> ToolResult {
        let path_text = input["path"].as_str().unwrap();
        let content = input["content"].as_str().unwrap();
        let create_if_missing = input
            .get("createIfMissing")
            .and_then(Value::as_bool)
            .unwrap_or(true);
        let expected_hash = input.get("expectedSha256").and_then(Value::as_str);
        let resolved = match resolve_writable(Path::new(path_text), &context.cwd).await {
            Ok(path) => path,
            Err(error) => {
                let code = if error.message.contains("超出工作区") {
                    "PATH_OUTSIDE_WORKSPACE"
                } else {
                    "FILE_WRITE_ERROR"
                };
                return ToolResult::failure(
                    code,
                    error.message,
                    context.started_at,
                    json!({"path":path_text}),
                );
            }
        };
        let old_bytes = match fs::read(&resolved).await {
            Ok(bytes) => Some(bytes),
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {
                if !create_if_missing {
                    return ToolResult::failure(
                        "FILE_NOT_FOUND",
                        format!("文件不存在且 createIfMissing 为 false：{path_text}"),
                        context.started_at,
                        json!({"path":path_text}),
                    );
                }
                None
            }
            Err(error) => {
                return ToolResult::failure(
                    "FILE_WRITE_ERROR",
                    error.to_string(),
                    context.started_at,
                    json!({"path":path_text}),
                );
            }
        };
        let pre_hash = old_bytes.as_deref().map(sha256);
        if let Some(expected_hash) = expected_hash
            && pre_hash.as_deref() != Some(expected_hash)
        {
            return ToolResult::failure(
                "HASH_MISMATCH",
                format!(
                    "文件哈希不匹配。预期 {expected_hash}，实际 {}。",
                    pre_hash.as_deref().unwrap_or("<missing>")
                ),
                context.started_at,
                json!({"path":path_text,"expectedSha256":expected_hash,"actualSha256":pre_hash}),
            );
        }
        let Some(parent) = resolved.parent() else {
            return ToolResult::failure(
                "FILE_WRITE_ERROR",
                "目标文件缺少父目录。",
                context.started_at,
                json!({"path":path_text}),
            );
        };
        if let Err(error) = fs::create_dir_all(parent).await {
            return ToolResult::failure(
                "FILE_WRITE_ERROR",
                error.to_string(),
                context.started_at,
                json!({"path":path_text}),
            );
        }
        let tmp = resolved.with_extension(format!("xycli-tmp-{}", Uuid::new_v4()));
        let write_result = async {
            fs::write(&tmp, content.as_bytes()).await?;
            fs::rename(&tmp, &resolved).await
        }
        .await;
        if let Err(error) = write_result {
            let _ = fs::remove_file(&tmp).await;
            return ToolResult::failure(
                "FILE_WRITE_ERROR",
                error.to_string(),
                context.started_at,
                json!({"path":path_text}),
            );
        }
        let old_text = old_bytes.as_deref().map(String::from_utf8_lossy);
        ToolResult::success(
            json!({
                "path": path_text,
                "created": old_bytes.is_none(),
                "preImageSha256": pre_hash,
                "postImageSha256": sha256(content.as_bytes()),
                "unifiedDiff": unified_diff(path_text, old_text.as_deref(), content),
            }),
            context.started_at,
            json!({"resolvedPath":resolved,"contentLength":content.len()}),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn 差异包含删除和新增行() {
        let diff = unified_diff("a.txt", Some("old"), "new");
        assert!(diff.contains("-old"));
        assert!(diff.contains("+new"));
    }
}
